rspace/src/test/scala/coop/rchain/rspace/IStoreTests.scala

```diff
@@ -0,0 +1,169 @@
+        store.removeDatum(txn, key, 0)
+        store.getChannels(txn, hash) shouldBe empty
+      }
+      store.clear()
+    }
+  }
+
+  "putWaitingContinuation" should "put waiting continuation in a new channel" in withTestStore {
+    store =>
+      forAll("channel", "continuation") { (channel: String, pattern: String) =>
+        val key          = List(channel)
+        val patterns     = List(StringMatch(pattern))
+        val continuation = new StringsCaptor
+        val wc: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(patterns, continuation, false)
+
+        store.withTxn(store.createTxnWrite()) { txn =>
+          store.putWaitingContinuation(txn, key, wc)
+          store.getWaitingContinuation(txn, key) shouldBe List(wc)
+        }
+        store.clear()
+      }
+  }
+
+  it should "append continuation if channel already exists" in withTestStore { store =>
+    forAll("channel", "continuation") { (channel: String, pattern: String) =>
+      val key          = List(channel)
+      val patterns     = List(StringMatch(pattern))
+      val continuation = new StringsCaptor
+      val wc1: WaitingContinuation[Pattern, StringsCaptor] =
+        WaitingContinuation(patterns, continuation, false)
+
+      val wc2: WaitingContinuation[Pattern, StringsCaptor] =
+        WaitingContinuation(List(StringMatch(pattern + 2)), continuation, false)
+
+      store.withTxn(store.createTxnWrite()) { txn =>
+        store.putWaitingContinuation(txn, key, wc1)
+        store.putWaitingContinuation(txn, key, wc2)
+        store.getWaitingContinuation(txn, key) should contain theSameElementsAs List(wc1, wc2)
+      }
+      store.clear()
+    }
+  }
+
+  "removeWaitingContinuation" should "remove waiting continuation from index" ignore withTestStore {
+    store =>
+      forAll("channel", "continuation") { (channel: String, pattern: String) =>
+        val key          = List(channel)
+        val patterns     = List(StringMatch(pattern))
+        val continuation = new StringsCaptor
+        val wc1: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(patterns, continuation, false)
+        val wc2: WaitingContinuation[Pattern, StringsCaptor] =
+          WaitingContinuation(List(StringMatch(pattern + 2)), continuation, false)
+
+        store.withTxn(store.createTxnWrite()) { txn =>
+          store.putWaitingContinuation(txn, key, wc1)
+          store.putWaitingContinuation(txn, key, wc2)
+          store.removeWaitingContinuation(txn, key, 0)
+          store.getWaitingContinuation(txn, key) shouldBe List(wc1)
+        }
+        store.clear()
+      }
+  }
+
+  "addJoin" should "add join for a channel" ignore withTestStore { store =>
+    forAll("channel", "channels") { (channel: String, channels: List[String]) =>
+      store.withTxn(store.createTxnWrite()) { txn =>
+        store.addJoin(txn, channel, channels)
+        store.getJoin(txn, channel) shouldBe List(channels)
```

Comment:
Similarly here, the results are different (hence `ignore`). Should I take `LMDBStore` as reference and modify `InMemoryStore` accordingly? @henrytill @ys-pyrofex

Reply:
Yes, but this change will be a bit more invasive.

To bring the implementations in alignment:
```diff
diff --git a/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala b/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
index 1cd465e2..65777c24 100644
--- a/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
+++ b/rspace/src/test/scala/coop/rchain/rspace/test/InMemoryStore.scala
@@ -17,7 +17,7 @@ class InMemoryStore[C, P, A, K <: Serializable] private (
     _keys: mutable.HashMap[String, Seq[C]],
     _waitingContinuations: mutable.HashMap[String, Seq[WaitingContinuation[P, K]]],
     _data: mutable.HashMap[String, Seq[Datum[A]]],
-    _joinMap: mutable.MultiMap[C, String],
+    _joinMap: mutable.HashMap[C, Seq[String]],
 )(implicit sc: Serialize[C])
     extends IStore[C, P, A, K]
     with ITestableStore[C, P] {
```

Then `addJoin` and `removeJoin` will need to rewritten to match the `LMDBStore` implementation.

I've broken that work out as a separate ticket here:
https://rchain.atlassian.net/browse/CORE-558

I've assigned it to you @lukasz-golebiewski, but if you want to pass it on to @dzajkowski as a warm-up, that would also be cool.